lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,577 @@
+		return
+	}
+	if !isSubpath(evPath, w.folderPath) {
+		panic(fmt.Sprintf("bug: Detected change outside of root directory for folder %v at %s", w.folderCfg.Description(), evPath))
+	}
+	relPath, _ := filepath.Rel(w.folderPath, evPath)
+	if w.pathInProgress(relPath) {
+		l.Debugf("%v Skipping path we modified: %s", w, relPath)
+		return
+	}
+	if w.folderIgnores.ShouldIgnore(relPath) {
+		l.Debugf("%v Ignoring: %s", w, relPath)
+		return
+	}
+	w.aggregateEvent(relPath, time.Now(), evType)
+}
+
+// Provide to be checked path first, then the path of the folder root.
+var isSubpath = strings.HasPrefix
+
+func (w *watcher) resetNotifyTimerIfNeeded() {
+	if w.notifyTimerNeedsReset {
+		w.resetNotifyTimer(w.notifyDelay)
+	}
+}
+
+// resetNotifyTimer should only ever be called when notifyTimer has stopped
+// and notifyTimer.C been read from. Otherwise, call resetNotifyTimerIfNeeded.
+func (w *watcher) resetNotifyTimer(duration time.Duration) {
+	l.Debugf("%v Resetting notifyTimer to %s", w, duration.String())
+	w.notifyTimerNeedsReset = false
+	w.notifyTimer.Reset(duration)
+}
+
+func (w *watcher) aggregateEvent(evPath string, evTime time.Time, evType eventType) {
+	if evPath == "." || w.rootEventDir.eventCount() == maxFiles {
+		l.Debugln(w, "Scan entire folder")
+		firstModTime := evTime
+		if w.rootEventDir.childCount() != 0 {
+			evType |= w.rootEventDir.eventType()
+			firstModTime = w.rootEventDir.firstModTime()
+		}
+		w.rootEventDir = newEventDir(".", nil)
+		w.rootEventDir.events["."] = &event{
+			path:         ".",
+			firstModTime: firstModTime,
+			lastModTime:  evTime,
+			evType:       evType,
+		}
+		w.resetNotifyTimerIfNeeded()
+		return
+	}
+
+	parentDir := w.rootEventDir
+
+	// Check if any parent directory is already tracked or will exceed
+	// events per directory limit bottom up
+	pathSegments := strings.Split(filepath.ToSlash(evPath), "/")
+
+	// As root dir cannot be further aggregated, allow up to maxFiles
+	// children.
+	localMaxFilesPerDir := maxFiles
+	var currPath string
+	for i, pathSegment := range pathSegments[:len(pathSegments)-1] {
+		currPath = filepath.Join(currPath, pathSegment)
+
+		if event, ok := parentDir.events[currPath]; ok {
+			event.lastModTime = evTime
+			event.evType |= evType
+			l.Debugf("%v Parent %s (type %s) already tracked: %s", w, currPath, event.evType, evPath)
+			return
+		}
+
+		if parentDir.childCount() == localMaxFilesPerDir {
+			l.Debugf("%v Parent dir %s already has %d children, tracking it instead: %s", w, currPath, localMaxFilesPerDir, evPath)
+			w.aggregateEvent(filepath.Dir(currPath),
+				evTime, evType)
+			return
+		}
+
+		// If there are no events below path, but we need to recurse
+		// into that path, create eventDir at path.
+		if _, ok := parentDir.dirs[currPath]; !ok {
+			l.Debugf("%v Creating eventDir: %s", w, currPath)
+			parentDir.dirs[currPath] = newEventDir(currPath, parentDir)
+		}
+		parentDir = parentDir.dirs[currPath]
```

Comment:
Could be rewritten as:


```
if newParent, ok := parentDir.dirs[currPath]; ok {
    parentDir = newParent
} else {
    parentDir = parentDir.dirs[currPath] = newEventDir(currPath, parentDir)
}
```

to minimize map access to 2 in worst case, instead of 3, and 1 in best case instead of 2.